File: source/go/internal/core/vara.go
package core

import (
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/types"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/utils"
)

func ComputeVara(dateUtcMs, sunriseUtcMs int64, varaNames [7]types.VaraName) types.VaraInfo {
	d := dateUtcMs
	if dateUtcMs < sunriseUtcMs {
		d = dateUtcMs - 86_400_000
	}
	index := types.Date(d).UTCDay()
	return types.VaraInfo{
		Index:       index,
		Name:        varaNames[index].Name,
		ShortName:   varaNames[index].Short,
		EnglishName: utils.EnglishDayNames[index],
	}
}

File: source/ts/src/core/vara.ts
import type { VaraInfo } from '../types/elements';
import { ENGLISH_DAY_NAMES } from '../utils/constants';

/** The Hindu day starts at SUNRISE: an instant before it belongs to the previous day's Vara. */
export function computeVara(
  dateUtc: Date,
  sunriseUtc: Date,
  varaNames: readonly { name: string; short: string }[],
): VaraInfo {
  let d = new Date(dateUtc);
  if (dateUtc.getTime() < sunriseUtc.getTime()) {
    d = new Date(dateUtc.getTime() - 86_400_000);
  }
  const index = d.getUTCDay();
  return {
    index,
    name: varaNames[index]!.name,
    shortName: varaNames[index]!.short,
    englishName: ENGLISH_DAY_NAMES[index]!,
  };
}
